pkg/distributor/query.go

```diff
@@ -204,11 +205,11 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 	queryLimiter := limiter.QueryLimiterFromContextWithFallback(ctx)
 	reqStats := stats.FromContext(ctx)
 
-	queryIngester := func(ctx context.Context, ing *ring.InstanceDesc, cancelContext context.CancelFunc) (ingesterQueryResult, error) {
+	queryIngester := func(ctx context.Context, ing *ring.InstanceDesc, cancelContext context.CancelCauseFunc) (ingesterQueryResult, error) {
 		log, ctx := spanlogger.NewWithLogger(ctx, d.log, "Distributor.queryIngesterStream")
 		cleanup := func() {
 			log.Span.Finish()
-			cancelContext()
+			cancelContext(cancellation.NewErrorf("stream closed"))
```

Comment:
would it help debugging if we included the returned error from `queryIngester`? Part of the problem is that that error won't always be present, especially when queryIngester returns before the cleanup happens, but it will help when debugging early returns

Reply:
I don't think that's necessary: there are only three kinds of errors that will be returned by `queryIngester`, and they'll be logged:

* limit errors will cause the query to be terminated, and these errors will be logged and added to the span [here](https://github.com/grafana/mimir/blob/aab75eb43a89a4ff5a066829d2c0d18e1ccbee0e/vendor/github.com/grafana/dskit/ring/replication_set.go#L345)
* issues communicating with the ingester (errors returned by `Recv()`) will be added to the span by the gRPC tracing instrumentation [here](https://github.com/grafana/mimir/blob/aab75eb43a89a4ff5a066829d2c0d18e1ccbee0e/vendor/github.com/opentracing-contrib/go-grpc/client.go#L146) (but not logged), but will be logged and added to the span [here](https://github.com/grafana/mimir/blob/aab75eb43a89a4ff5a066829d2c0d18e1ccbee0e/vendor/github.com/grafana/dskit/ring/replication_set_tracker.go#L302) if they're the first failure for a zone
* errors getting the gRPC client or calling `QueryStream()` will be logged and added to the span [here](https://github.com/grafana/mimir/blob/aab75eb43a89a4ff5a066829d2c0d18e1ccbee0e/vendor/github.com/grafana/dskit/ring/replication_set_tracker.go#L302) if the error is the first failure for a zone

Reply:
makes sense. What we have is probably already sufficient. Thanks for explaining